### PR TITLE
GRANITE-29232: oak-search-elastic review/update

### PR DESCRIPTION
**Pull request by fabriziofortino**
This PR contains a lot of stuff in order to update the elastic module. Have a look at the single commits to have a better context of the changes.

### CODE REVIEW

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+    @Rule
+    public final TemporaryFolder folder = new TemporaryFolder(new File("target"));
+
+    @Rule
+    public final OsgiContext context = new OsgiContext();
+
+    private final ElasticsearchIndexProviderService service = new ElasticsearchIndexProviderService();
+
+    private MountInfoProvider mip;
```
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderServiceTest.java]: Codacy Issue found: [Perhaps 'mip' could be replaced by a local variable.](https://app.codacy.com/gh/oak-indexing/jackrabbit-oak/pullRequest?prid=5120315)

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    private Whiteboard wb;
```
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderServiceTest.java]: Codacy Issue found: [Avoid unused private fields such as 'wb'.](https://app.codacy.com/gh/oak-indexing/jackrabbit-oak/pullRequest?prid=5120315)

Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderServiceTest.java]: Codacy Issue found: [Perhaps 'wb' could be replaced by a local variable.](https://app.codacy.com/gh/oak-indexing/jackrabbit-oak/pullRequest?prid=5120315)

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderServiceTest.java (hunk 1 above)
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderServiceTest.java]: Codacy Issue found: [Perhaps 'mip' could be replaced by a local variable.](https://app.codacy.com/gh/oak-indexing/jackrabbit-oak/pullRequest?prid=5120315)

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderServiceTest.java (hunk 2 above)
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderServiceTest.java]: Codacy Issue found: [Avoid unused private fields such as 'wb'.](https://app.codacy.com/gh/oak-indexing/jackrabbit-oak/pullRequest?prid=5120315)

Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderServiceTest.java]: Codacy Issue found: [Perhaps 'wb' could be replaced by a local variable.](https://app.codacy.com/gh/oak-indexing/jackrabbit-oak/pullRequest?prid=5120315)

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderServiceTest.java (hunk 1 above)
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderServiceTest.java]: Codacy Issue found: [Perhaps 'mip' could be replaced by a local variable.](https://app.codacy.com/gh/oak-indexing/jackrabbit-oak/pullRequest?prid=5120315)

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderServiceTest.java (hunk 2 above)
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderServiceTest.java:51]: Codacy Issue found: [Avoid unused private fields such as 'wb'.](https://app.codacy.com/gh/oak-indexing/jackrabbit-oak/pullRequest?prid=5120315)

Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderServiceTest.java:51]: Codacy Issue found: [Perhaps 'wb' could be replaced by a local variable.](https://app.codacy.com/gh/oak-indexing/jackrabbit-oak/pullRequest?prid=5120315)

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchPropertyIndexTest.java [hunk 3]
```diff
@@ -34,29 +34,33 @@
...
 import static org.apache.jackrabbit.oak.plugins.index.search.FulltextIndexConstants.PROPDEF_PROP_NODE_NAME;
 import static org.hamcrest.CoreMatchers.containsString;
 import static org.hamcrest.MatcherAssert.assertThat;
 
 public class ElasticsearchPropertyIndexTest extends AbstractQueryTest {
-    @Rule
-    public ElasticsearchManagementRule esMgmt = new ElasticsearchManagementRule();
+    @ClassRule
+    public static final ElasticsearchContainer ELASTIC =
+            new ElasticsearchContainer("docker.elastic.co/elasticsearch/elasticsearch:7.6.0");
```
Comment (nit0906) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchPropertyIndexTest.java]: @fabriziofortino 
This would expect docker service to be available on the instance running this test - correct ? Might be problematic in case of windows (I face a lot of issues while running docker on windows).

Maybe we should have an option where this test can be run with an external elastic server as well ? Maybe system properties - if they are present, they take precedence otherwise the docker container is used. wdyt ?

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/index/ElasticsearchIndexEditorProvider.java [hunk 4]
```diff
@@ -49,15 +48,15 @@ public ElasticsearchIndexEditorProvider(@NotNull ElasticsearchIndexCoordinateFac
                                            @NotNull NodeBuilder definition, @NotNull NodeState root,
                                            @NotNull IndexUpdateCallback callback) throws CommitFailedException {
         if (TYPE_ELASTICSEARCH.equals(type)) {
-            checkArgument(callback instanceof ContextAwareCallback, "callback instance not of type " +
-                    "ContextAwareCallback [%s]", callback);
-            IndexingContext indexingContext = ((ContextAwareCallback)callback).getIndexingContext();
+            assert callback instanceof ContextAwareCallback : "callback instance not of type " +
```
Comment (nit0906) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/index/ElasticsearchIndexEditorProvider.java]: I am not very sure about this - but I think assertions disabled by default (like you need to pass -ea in the java args to enable them).

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchPropertyIndexTest.java (hunk 3 above)
Comment (fabriziofortino) [reply oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchPropertyIndexTest.java]: To do that we should remove the `ClassRule` and manage the lifecycle for each test. Out of curiosity: what kind of issues do you have running it on docker for windows?

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchClientFactory.java [hunk 5]
```diff
@@ -25,32 +24,36 @@
...
-import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
 
-public class ElasticsearchConnectionFactory implements Closeable {
-    private static final Logger LOG = LoggerFactory.getLogger(ElasticsearchConnectionFactory.class);
-    private final ConcurrentMap<ElasticsearchCoordinate, RestHighLevelClient> clientMap = Maps.newConcurrentMap();
+public class ElasticsearchClientFactory implements Closeable {
+    private static final Logger LOG = LoggerFactory.getLogger(ElasticsearchClientFactory.class);
 
+    public static final ElasticsearchClientFactory INSTANCE = new ElasticsearchClientFactory();
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchClientFactory.java]: Does it need to be public, if getInstance is public?

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchClientFactory.java [hunk 6]
```diff
@@ -61,7 +64,6 @@ public RestHighLevelClient getConnection(ElasticsearchCoordinate esCoord) {
     public void close() {
         lock.writeLock().lock();
         try {
-            isClosed.set(true);
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchClientFactory.java]: Usually it's a good idea to try avoiding to close twice... why did you remove it?

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchCoordinate.java [hunk 7]
```diff
@@ -16,18 +16,85 @@
...
+    }
+
+    public static ElasticsearchCoordinate build(@NotNull Map<String, ?> config) {
+        ElasticsearchCoordinate coordinate = null;
+        Object p = config.get(PORT_PROP);
+        if (p != null) {
+            try {
+                Integer port = Integer.parseInt(p.toString());
+                coordinate = build((String) config.get(SCHEME_PROP), (String) config.get(HOST_PROP), port);
+            } catch (NumberFormatException nfe) { /* ignore */ }
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchCoordinate.java]: I wouldn't ignore, but instead log a warning.

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchCoordinate.java [hunk 8]
```diff
@@ -16,18 +16,85 @@
...
+        }
+        return coordinate;
+    }
+
+    private static ElasticsearchCoordinate build(String scheme, String host, Integer port) {
+        if (scheme == null || host == null || port == null) {
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchCoordinate.java]: I would probably throw a NPE if one is null

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderService.java [hunk 9]
```diff
@@ -209,45 +208,51 @@ private void initializeExtractedTextCache(Map<String, ?> config, StatisticsProvi
...
+        ElasticsearchCoordinate coordinate = ElasticsearchCoordinate.build(System.getProperties().entrySet()
+                .stream()
+                .collect(Collectors.toMap(
+                        e -> String.valueOf(e.getKey()),
+                        e -> String.valueOf(e.getValue()))
+                )
+        );
+
+        if (coordinate == null) {
+            coordinate = ElasticsearchCoordinate.build(contextConfig);
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderService.java]: I see now why you return null in build... but I think it's better to have the logic in the caller than in build.

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchPropertyIndexTest.java (hunk 3 above)
Comment (nit0906) [reply oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchPropertyIndexTest.java]: > To do that we should remove the ClassRule and manage the lifecycle for each test

Yes, but I understand that it would be a bit of an over work for a test - maybe we can let it be the way it is currently and add a TODO to maybe have a look at it later. 


> what kind of issues do you have running it on docker for windows?

For instance, I had configured it as a startup service, but somehow it never did. And even after manually enabling the service, It used to get stopped every now and then..I gave up trying to fix it after a while and moved on to a linux VM for docker related stuff. (But this was a bit back, probably there could be better support for windows now)

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchClientFactory.java (hunk 5 above)
Comment (fabriziofortino) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchClientFactory.java]: done

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchClientFactory.java (hunk 6 above)
Comment (fabriziofortino) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchClientFactory.java]: I had to remove it because the `getConnection` could not be used anymore when closed (see the throw Exception I removed). This was a problem in both the tests and also when the OSGI bundle gets deactivated and then re-activated.
Looking again at the code I figured it out the concurrent map needs to be cleared on close to avoid returning already closed clients: I have changed the code.

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchClientFactory.java [hunk 10]
```diff
@@ -25,32 +24,36 @@
...
-import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
 
-public class ElasticsearchConnectionFactory implements Closeable {
-    private static final Logger LOG = LoggerFactory.getLogger(ElasticsearchConnectionFactory.class);
-    private final ConcurrentMap<ElasticsearchCoordinate, RestHighLevelClient> clientMap = Maps.newConcurrentMap();
+public class ElasticsearchClientFactory implements Closeable {
+    private static final Logger LOG = LoggerFactory.getLogger(ElasticsearchClientFactory.class);
 
+    private static final ElasticsearchClientFactory INSTANCE = new ElasticsearchClientFactory();
```
Comment (tihom88) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchClientFactory.java]: Just thinking, should we create (initialize) this on demand?

**Review by nit0906 — APPROVED** [review]
LGTM

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchClientFactory.java [hunk 11]
```diff
@@ -25,32 +24,36 @@
...
+
+    private final ConcurrentMap<ElasticsearchCoordinate, RestHighLevelClient> clientMap = new ConcurrentHashMap<>();
```
Comment (tihom88) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchClientFactory.java]: Are we supporting multiple elastic servers?

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchClientFactory.java (hunk 6 above)
Comment (tihom88) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchClientFactory.java]: Here we are keeping all clients in a map and in a multi threaded environment if a thread call close we will be closing all and clearing complete map. Shouldn't we close connections opened by the specific thread only?

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchConnection.java [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+    protected static final int DEFAULT_PORT = 9200;
+
+    protected static final Supplier<ElasticsearchConnection> defaultConnection = () ->
+            new ElasticsearchConnection(DEFAULT_SCHEME, DEFAULT_HOST, DEFAULT_PORT);
+
+    private String scheme;
+    private String host;
+    private int port;
+
+    private volatile RestHighLevelClient client;
```
Comment (tihom88) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchConnection.java:54]: Can this be non volatile as each thread will be accessing it and this won't be changing frequently.

Comment (fabriziofortino) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchConnection.java:54]: It's better to keep it volatile because of the double-check synchronized in `getClient`.

**Comment by fabriziofortino** [comment]
merged with https://issues.apache.org/jira/browse/OAK-8965
